keps/0013-external-specs.md

```diff
@@ -0,0 +1,162 @@
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+## Summary
+
+Organizations and companies already spend time building, updating and debugging specs for deploying their applications. Asking them to spend additional time to maintain a KUDO spec may prevent adoption of the platform. Improving KUDO to be able to consume more than just a KUDO Spec defined in [KEP-0009](keps/0009-operator-toolkit.md) would allow for oragnizations to provide access to their applications to the KUDO community.
+
+## Motivation
+
+While being able to package our own application definition via [KEP0009](keps/0009-operator-toolkit.md), being able to leverage other application package formats for customization and extension would provide access to a trove of KUDO Frameworks without any additional work required by FrameworkVendors. There are
+
+- Frameworks
+- Helm
+- CNAB
+- OperatorHub
+
+### Goals
+
+- Run non-KUDO defined applications as Frameworks
+
+### Non-Goals
+
+- (currently) Manage repos of non-KUDO applications (e.g. Helm Charts)
+- Convert non-KUDO applications into KUDO format
+- Create a plugin system to upload new templating engines. Must be present in core KUDO codebase (for now)
+- Adjusting the capabilities of Parameters to support non-string values
+
+## Proposal
+
+### Engine Interface
+
+The core functionality required by KUDO is the ability to turn
+
+```go
+type StepArgument struct{
+  // The FrameworkVersion object to be used by the Engine
+  FrameworkVersion FrameworkVersion
+  // The Plan being executed
+  Plan Plan
+  // The Phase being executed
+  Phase Phase
+  // The step that is being currently run
+  Step Step
+  // The parameter values from the Instance Spec
+  Arguments map[string]string
+}
+```
+
+into a list of Kubernetes objects. Therefore, defining an `Engine` interface for a new application spec as follows would allow for using that spec in the execution of a plan.
+
+```go
+type Engine interface{
+  // Turns the input variables into a list of Kubernetes objects to apply
+  // for this step
+  Render(args StepArgument) ([]runtime.Objects, error)
+}
+```
+
+The `PlanExecution` controller would select the proper engine based on the `engine` keyword added to the `FrameworkVersion` spec.
+
+### Engine Spec in FrameworkVersion
+
+The Framework Version Spec would have a new field `Engine` that would specify the engine that should be used to Render steps into Kubernetes objects. Initial implementation will focus on adding one additional engine to support the execution of `Helm` charts.
```

Comment:
can we really do just one? What about various helm versions? Do we expect that new helm library will be always able to process old templates? Is it true for v2 and v3?

Reply:
> I must say I like this idea of extensions!
> 
> I am still trying to picture how will that look like from the user perspective though, can you maybe picture an example?
> 
> So I have mysql chart, I extend it, add plan and submit it. What I end up with?
> I will have 1 framework: Mysql

This goes to the outstanding questions:  Does a Framework Extension always fall within the base Framework or not?  I could see a case for each, but I imagine there would be edge cases where the extension isn't compatible (for an acceptable definition) with the base.

> I will have 2 frameworkVersions?: mysql- and 

There would be two FrameworkVersion, but both may not be installed on the cluster.  See the Outstanding Questions section.  Maybe the Extension FrameworkVersion captures all the required base data in its definition and the base is not required anywhere at runtime.   I would prefer not having to have the base installed on the cluster.  This will prevent someone from providing a custom extension to their customers, and not having to specify on the cluster which version of the Framework is the extension and which is the base.

> I will have 1 instance created from the second FV?
Any instance will be an instance of whatever FV you specify.

Reply:
I was told by someone on the helm team that they should be backwards compatible.  i.e. helm 3 can render helm 2.